rfcs/0031-bazel.md

```diff
@@ -0,0 +1,358 @@
+| 14 | all | go-libp2p | 4 | W 5/27  |
+| 15 | all | CUDA code in Coda | 1 | TBD | cuda-fixnum, cuda-prover, etc. |
+
+Note: I don't have a GPU, and I'm not sure if I can compile CUDA code
+without one.  If I can, task 15 will be completed by end of month.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+The learning curve for Bazel is little steep in places, plus it has
+its own languge, Starlark - yet another language to learn.  It's a
+Python dialect, but still, somebody on staff will probably have to be
+tasked with learning the stuff.  On the other hand, once the build
+system is in place it will probably not require much further work,
+just like Makefiles are usually left alone once they stabilize.
+
+Another issue with Bazel is feature detection.  Bazel just builds
+stuff; it does not provide the feature-detection functionality
+supported by GNU autoconf tools.  So it has no analog to
+`./configure`; you have to explicitly tell it about all features.
+That means that a Bazel build, in cases where feature detection is
+required, must be complimented by a traditional `configure` file or
+something similar.
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Descriptions of the general benefits of Bazel are easily found with a
+simple websearch, so I'll keep it brief:
+* Reliable, replicable builds.
+* Support for multiple languages.
+* Developed and heavily used by Google; also used by many other major projects.
+* Eliminates the need to separately install/config/build third-party
+  dependencies (apt-get, homebrew, etc.) Bazel will download and build
+  everything.
+* Support for remote caching and execution is (probably) unique.
+
+Alternatives
+* Stick with current situation.  That might work for o(1) internally
+  but is not so attractive for outside volunteers who would have to
+  struggle with the build system.  My guess is that improving the
+  current system would probably be more expensive in the long run than
+  switching to Bazel.
+* Nix may be an alternative, but according to the folks who provide
+  Bazel support for Haskell, Bazel and Nix work well together: [NIX +
+  BAZEL = FULLY REPRODUCIBLE, INCREMENTAL
+  BUILDS](https://www.tweag.io/posts/2018-03-15-bazel-nix.html).
+
+## Prior art
+[prior-art]: #prior-art
+
+Blogposts about switching to Bazel can be found with a little web searching.
+
+I spent considerable effort a few years ago converting a modestly
+sized but somewhat complex C/C++ project
+([OpenOCF](https://github.com/OpenOCF/iochibity)) from SCons to Bazel.
+Though the project was never completed, the switch to Bazel was, and
+was (in my opinion) quite successful.
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* What parts of the design do you expect to resolve through the RFC process before this gets merged?
+
+  * Which platforms and toolchains are to be officially supported.
```

Comment:
In the short-term:

Linux and MacOS.

We should target x86_64 (and taking advantage of whichever SSE/AVX extensions we have turned on in the current CMake invocation for snarky). 

It would be preferable to support a Bazel toolchain via Nix for reproducibility's sake (as explained in https://www.tweag.io/posts/2018-03-15-bazel-nix.html ). But we should do this as a follow-up.

Initially, for macOS, supporting the "default" Xcode toolchain + OCaml installed in whichever way is easiest would work. For Linux, a Debian based Docker container installing the toolchain will suffice (similar to our existing `Dockerfile-toolchain` pulling from `ocaml/opam2:debian-9-ocaml-4.08`).